firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Util/Containers/RingBuffer.txt

```diff
@@ -0,0 +1,592 @@
+// * indicates the element addressed by oldest_index_
+// ^ indicates the element addressed by newest_index_
+
+SCENARIO('Non-Volatile RingBuffer works correctly when oldest index is 0') 
+  GIVEN('A RingBuffer with capacity 4 to which 0 bytes of data were pushed, 0 bytes were poped') 
+    // internal state should look like [*^] [] [] [] []
+    WHEN('1 byte of data is pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After push method ,the peek method returns expected data ') 
+        ////internal state should look like [0*] [^] [] [][]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('The pop method returns empty status')
+        // internal state should look like [*^] [] [] [][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 1 bytes of data were pushed and 0 byte was poped') 
+    // internal state should look like [0*] [^] [] [][]
+    WHEN('1 byte of data is pushed') 
+      THEN('The push method returns ok status') 
+      THEN('After push method, the peek method returns expected data ') 
+        // internal state should look like [0*] [1] [^] [][]
+
+    WHEN('1 byte of data is poped') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data') 
+        // internal state should look like [*^][][][][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 2 bytes of data were pushed and 0 byte were poped ') 
+    // internal state should look like [0*][1][^][][]
+    WHEN('1 byte of data is pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After push method, peek method returns expected data') 
+        // internal state should look like [0*][1][2][^][]
+
+    WHEN('1 byte of data is poped ') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data in RingBuffer') 
+        // internal state should look like [] [1*] [^] [][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 3 bytes of data were pushed and 0 byte were poped')   //[][][][3*][4^]
+    // internal state should look like [0*][1][2][^][]
+    WHEN('1 bytes of data are pushed ') 
+      THEN('The push method returns ok status') 
+      THEN('After the push method, peek method returns expected data') 
+        // internal state should look like [1*] [2] [3] [4][^]
+
+    WHEN('2 byte of data is poped ') 
+      THEN('The pop method returns ok status') 
+      THEN('After the pop method,the peek method returns expected data in RingBuffer') 
+        // internal state should look like [][][2*][^][]
+
+  GIVEN('A RingBuffer with capacity 4 to which 4 bytes of data were pushed and 0 byte were poped')   // [0^][][][][4*]
+    // internal state should look like [0*] [1] [2] [3][^]
+    WHEN('1 byte of data is pushed ') 
+      THEN('The push method returns full status') 
+      THEN('After the push method, peek method returns expected data') 
```

Comment:
It may be clearer to also explicitly say that the pushed byte of data is not in the RingBuffer object